### PR TITLE
Feature/caching strategies

### LINKED ISSUE #491: Caching strategies in View
- dont keep every resolved class (`Project.createFullView()`) - e.g. use LRU . maybe remove `Project.ondemandView()`  
- handling modifications - apply Modified MethodSource/ClassSource into the View

### PR DESCRIPTION
**Pull request by stschott**
Closes #491

### CODE REVIEW

File: sootup.core/src/main/java/sootup/core/cache/Cache.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package sootup.core.cache;
+
+import java.util.Collection;
+import javax.annotation.Nonnull;
+import sootup.core.model.SootClass;
+import sootup.core.types.ClassType;
+
+/** Interface for different caching strategies of resolved classes. */
+public interface Cache<S extends SootClass<?>> {
```
Comment (swissiety) [inline sootup.core/src/main/java/sootup/core/cache/Cache.java]: naming: ClassCache?

File: sootup.core/src/main/java/sootup/core/cache/LRUCache.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  @Override
+  public synchronized S getClass(ClassType classType) {
+    S sootClass = cache.get(classType);
+    if (sootClass != null) {
+      accessOrder.remove(classType);
+      accessOrder.addFirst(classType);
+    }
+
```
Comment (swissiety) [inline sootup.core/src/main/java/sootup/core/cache/LRUCache.java:28]: should we throw an exception?

File: sootup.core/src/main/java/sootup/core/cache/MutableCache.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package sootup.core.cache;
+
+import sootup.core.model.SootClass;
+import sootup.core.types.ClassType;
+
+/** Interface for caches which are mutable and allow classes to be removed from. */
+public interface MutableCache<S extends SootClass<?>> extends Cache<S> {
+  S removeClass(ClassType classType);
```
Comment (swissiety) [inline sootup.core/src/main/java/sootup/core/cache/MutableCache.java]:
```suggestion
S removeClass(ClassType classType);
default S modifyClass(@Nonnull ClassType classType){
      removeClass(classType);
      addClass(classType);
  }
```
as a convenience method?

File: sootup.java.core/src/main/java/sootup/java/core/views/JavaView.java [hunk 4]
```diff
@@ -154,14 +181,26 @@ protected synchronized Optional<JavaSootClass> buildClassFrom(
...
+        getProject().getInputLocations().stream()
+            .flatMap(location -> location.getClassSources(this).stream())
+            .map(this::buildClassFrom)
+            .collect(Collectors.toList());
+
+    Collection<JavaSootClass> resolvedClasses =
+        resolvedClassesOpts.stream()
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .collect(Collectors.toList());
```
Comment (swissiety) [inline sootup.java.core/src/main/java/sootup/java/core/views/JavaView.java:200]: please join/merge these streams to allocate the Collection just once

File: sootup.core/src/main/java/sootup/core/cache/MutableCache.java (hunk 3 above)
Comment (stschott) [reply sootup.core/src/main/java/sootup/core/cache/MutableCache.java]: Do you mean a replaceClass method, which removes an old class and instead adds a new one?
The method you are suggesting adds and removes the same class.

File: sootup.core/src/main/java/sootup/core/cache/LRUCache.java (hunk 2 above)
Comment (stschott) [reply sootup.core/src/main/java/sootup/core/cache/LRUCache.java:28]: I think we should stick to default Map behavior and return null, if the element is not present.

File: sootup.core/src/main/java/sootup/core/cache/MutableCache.java (hunk 3 above)
Comment (swissiety) [reply sootup.core/src/main/java/sootup/core/cache/MutableCache.java]: sure.. someone was excited of using the suggestion feature ;)

**Review by JonasKlauke — APPROVED** [review]

**Pull request merged**